src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,684 @@
+     *     the {@code KDFParameters} used to configure the derivation
+     *     algorithm or {@code null} if no parameters are provided
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified {@code KDF}
+     *     algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the specified provider does not support a {@code KDFSpi}
+     *     implementation for the specified algorithm and parameters
+     * @throws NullPointerException
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "algorithm must not be null");
+        Objects.requireNonNull(provider, "provider must not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF(new Delegate((KDFSpi) instance.impl,
+                                        instance.provider), algorithm, kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey} object.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param derivationSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return the derived key
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationSpec} is
+     *     invalid or if the combination of {@code alg} and the {@code derivationSpec}
+     *     results in something invalid
+     * @throws NoSuchAlgorithmException
+     *     if {@code alg} is empty or invalid
```

Comment:
Is it easy to tell precisely what falls into "combination of alg and derivationSpec results in something invalid" and "alg is invalid"?